File: chapt5/chapt5_ws/src/demo_python_tf/demo_python_tf/tf_broadcast.py
import rclpy
from rclpy.node import Node
from tf2_ros import TransformBroadcaster
from geometry_msgs.msg import TransformStamped
from geometry_msgs.msg import Transform
from tf_transformations import quaternion_from_euler
import math
class TFBroadcaster(Node):
    def __init__(self, node_name):
        super().__init__(node_name)
        self.tf_broadcaster_ = TransformBroadcaster(self)
        self.timer_ = self.create_timer(0.1, self.tf_callback)
    def publish_tf(self, frame_id, child_frame_id, transform):
        tf = TransformStamped()
        tf.header.frame_id = frame_id
        tf.child_frame_id = child_frame_id
        tf.header.stamp = self.get_clock().now().to_msg()
        tf.transform = transform
        self.tf_broadcaster_.sendTransform(tf)
        self.get_logger().info(f'Published static transform from {frame_id} to {child_frame_id}')
    def tf_callback(self):
        frame_id = 'map'
        child_frame_id = 'odom'
        transform = Transform()
        transform.translation.x = 1.0
        transform.translation.y = 1.0
        transform.translation.z = 0.0
        q = quaternion_from_euler(0.0, 0.0, 0.0)
        transform.rotation.x = q[0]
        transform.rotation.y = q[1]
        transform.rotation.z = q[2]
        transform.rotation.w = q[3]
        self.publish_tf(frame_id, child_frame_id, transform)
        

def main():
    rclpy.init()
    node = TFBroadcaster('tf_broadcaster')
    rclpy.spin(node)
    rclpy.shutdown()

File: chapt5/chapt5_ws/src/demo_cpp_tf/src/tf_broadcast.cpp
#include <geometry_msgs/msg/transform_stamped.hpp>
#include <rclcpp/rclcpp.hpp>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2_ros/transform_broadcaster.hpp>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
#include <chrono>
using namespace std::chrono_literals;
using TFB = tf2_ros::TransformBroadcaster;
class TFBroadcaster : public rclcpp::Node
{
public:
    TFBroadcaster(const std::string &node_name) : Node(node_name)
    {
        broadcaster_odom_in_map = std::make_shared<TFB>(this);
        timer_ = this->create_wall_timer(100ms, std::bind(&TFBroadcaster::publish_odom_in_map, this));
    }
    void publish_tf(std::shared_ptr<TFB> broadcaster, const std::string &parent_frame, const std::string &child_frame,
                    const double &x, const double &y, const double &z,
                    const double &roll, const double &pitch, const double &yaw)
    {
        geometry_msgs::msg::Transform tf;
        tf.translation.x = x;
        tf.translation.y = y;
        tf.translation.z = z;
        tf2::Quaternion q;
        q.setRPY(roll, pitch, yaw);
        tf.rotation = tf2::toMsg(q);
        geometry_msgs::msg::TransformStamped tfs;
        tfs.header.stamp = this->now();
        tfs.header.frame_id = parent_frame;
        tfs.child_frame_id = child_frame;
        tfs.transform = tf;
        broadcaster->sendTransform(tfs);
        RCLCPP_INFO(this->get_logger(), "Published tf %s in %s", child_frame.c_str(), parent_frame.c_str());
    }
    void publish_odom_in_map()
    {
        publish_tf(broadcaster_odom_in_map, "map", "odom",
                          1.0, 1.0, 1.0,
                          0.0, 0.0, 0.0);
    }
private:
    std::shared_ptr<TFB> broadcaster_odom_in_map;
    rclcpp::TimerBase::SharedPtr timer_;

};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    std::shared_ptr<TFBroadcaster> node = std::make_shared<TFBroadcaster>("tf_broadcaster");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}